src/main/java/com/amazonaws/kinesisvideo/demoapp/activity/WebRtcActivity.java

```diff
@@ -259,6 +266,47 @@ private boolean isValidClient() {
         return client != null && client.isOpen();
     }
 
+    private void handlePendingIceCandidates(String clientId) {
+        // Add any pending ICE candidates from the queue for the client ID
+        Queue<IceCandidate> pendingIceCandidatesQueueByClientId = pendingIceCandidatesMap.get(clientId);
+        while(!pendingIceCandidatesQueueByClientId.isEmpty()) {
+            final IceCandidate iceCandidate = pendingIceCandidatesQueueByClientId.peek();
+            final boolean addIce = localPeer.addIceCandidate(iceCandidate);
+            Log.d(TAG, "Added ice candidate after SDP exchange " + iceCandidate + " " + (addIce ? "Successfully" : "Failed"));
+          pendingIceCandidatesQueueByClientId.remove();
+        }
+        // After sending pending ICE candidates, the client ID's peer connection need not be tracked
+        pendingIceCandidatesMap.remove(clientId);
+        peerConnectionFoundMap.remove(clientId);
+    }
+
+    private void checkAndAddIceCandidate(Event message, IceCandidate iceCandidate) {
+        // if answer/offer is not received, it means peer connection is not found. Hold the received ICE candidates in the map.
+        if(!peerConnectionFoundMap.containsKey(message.getSenderClientId())) {
+            Log.d(TAG, "SDP exchange is not complete. Ice candidate " + iceCandidate + " + added to pending queue");
+            // Add ICE candidate to a queue
+            peerConnectionFoundMap.put(message.getSenderClientId(), false);
+            Queue<IceCandidate> pendingIceCandidatesQueueByClientId = new LinkedList<>();
+            pendingIceCandidatesQueueByClientId.add(iceCandidate);
+            pendingIceCandidatesMap.put(message.getSenderClientId(), pendingIceCandidatesQueueByClientId);
+        }
+        // If the entry for client ID is available in the map but peer connection is not found, just update the ICE candidate map
+        // This would happen if we get subsequent ICE candidates before peer connection is established
+        else if (!peerConnectionFoundMap.get(message.getSenderClientId())) {
```

Comment:
You might want to re-structure this function after we discussed the changes. Per comments.

1) When receiving ICE candidate check the hashmap of active sessions by client id. 
2) If the client id has an entry to a peer connection then this is an IN-ORDER message and you can simply give the ICE candidate to the peer. DONE.
3) This is an OUT-OF-ORDER message (ICE candidate). Check if the clientID -> message queue hash map contains the entry
4) If it doesn't, create an entry (create the queue). Add the ICE candidate to the queue. This is the first ICE candidate in the out-of-order message.
5) If the entry exists, add the ICE candidate to the already existing message queue.
6) Need to add a timer of some sort that monitors the pending message hash map and removes the entries when those have been over 20 seconds.

Reply:
As discussed, we can handle the timer as a separate work item.